### DISCUSSION TITLE
Reactive Method interceptor

### DISCUSSION THREAD

**Discussion by tinyzero4** [quarkusio/quarkus#30099] [Community]
Hi guys, 
Do Quarkus support reactive interceptor in any way via Uni/Coroutines ?

Let's say we need to implement a custom rate limiter, so for specific methods we should wrap execution via rate limiter interceptor. However it turns out it is not possible to do this via kotlin coroutines

```
@RateLimiter
@Priority(2022)
@Interceptor
class RateLimiterInterceptor(private val client: RedissonReactiveClient) {
    @Inject
    private var route: RoutingContext? = null

    @AroundInvoke

    suspend fun invoke(context: InvocationContext): Any? {
        val definition = context.method.getAnnotation(RateLimiter::class.java)
        val user = resolveUser(context, route)

        check(definition, user)

        return context.proceed()
    }
    
    private suspend check(definition: RateLimiter, user: UserInfo) {
          // Request data from redis
    }
}
```

On build step i get the following error:

`[error]: Build step io.quarkus.arc.deployment.ArcProcessor#registerBeans threw an exception: java.lang.IllegalStateException: An interceptor method must accept exactly one parameter of type javax.interceptor.InvocationContext: java.lang.Object invoke(javax.interceptor.InvocationContext, kotlin.coroutines.Continuation<java.lang.Object>) declared on`

Is it possible somehow to implement method interceptor in a reactive way? Maybe return Uni, Mono, or smth else?

**Comment by mkouba** [comment]
> Is it possible somehow to implement method interceptor in a reactive way? Maybe return Uni, Mono, or smth else?

Well, if an _intercepted method_ returns `Uni` (or an async type in general; and this can be enforced/validated at runtime or even build time) then you can simply wrap the returned `Uni` (i.e. the result of `InvocationContext#proceed()`) and return a modified reactive stream instead. Otherwise you'd need to block.

**Reply by ichrist-gr** [reply]
Could you please provide an impementation example, because in my case the interceptor is executed after the jax-rs resource method. I'm using resteasy reactive jackson.

**Reply by alex-martel** [reply]
@ichrist-gr  take a look at this comment https://github.com/quarkusio/quarkus/discussions/28637#discussioncomment-6421786 which contains a sample app. hope that helps.
